File: lintcode/binary_tree_level_order_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Level order in a list of lists of integers
    """
    def levelOrder(self, root):
        def helper(root, level, levels):
            if root is not None:
                if len(levels) <= level:
                    levels.append([])
                levels[level].append(root.val)
                helper(root.left, level + 1, levels)
                helper(root.right, level + 1, levels)
            return levels
        return helper(root, 0, [])

File: leetcode/binary_tree_level_order_traversal.js
/* Given a binary tree, return the level order traversal of its nodes' values. 
 * (ie, from left to right, level by level).
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    function helper(root, level, levels) {
        if (root) {
            if (levels.length <= level)
                levels.push([]);
            levels[level].push(root.val);
            helper(root.left, level + 1, levels);
            helper(root.right, level + 1, levels);
        }
        return levels;
    }
    return helper(root, 0, []);
};
